packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js

```diff
@@ -0,0 +1,154 @@
+// with custom symbols. Ex: "USDETH * COMPUSD * MY_CUSTOM_FEED".
+class ExpressionPriceFeed extends PriceFeedInterface {
+  /**
+   * @notice Constructs the DominationFinancePriceFeed.
+   * @param {Object} priceFeedMap an object mapping from price feed names to the price feed objects.
+   *                 Ex:
+   *                 {
+   *                   "USDETH": ETHBTC_PRICE_FEED_INSTANCE,
+   *                   "COMPUSD": COMPUSD_PRICE_FEED_INSTANCE
+   *                 }
+   * @param {string} expression a string expression that uses price feeds in the priceFeedMap to compute a price.
+   *                 Note: all symbols used in this expression must be defined in the price feed map.
+   *                 Ex: "(USDETH + COMPUSD) / COMPUSD"
+   * @param {number} decimals decimals to use in the price output.
+   */
+  constructor(priceFeedMap, expression, decimals = 18) {
+    super();
+    this.expressionCode = math.parse(expression).compile();
+    this.priceFeedMap = priceFeedMap;
+    this.decimals = decimals;
+  }
+
+  async getHistoricalPrice(time) {
+    const historicalPrices = {};
+    const errors = [];
+    await Promise.all(
+      Object.entries(this.priceFeedMap).map(async ([name, pf]) => {
+        const price = await pf.getHistoricalPrice(time).catch(err => errors.push(err));
+        historicalPrices[name] = this._convertToDecimal(price, pf.getPriceFeedDecimals());
+      })
+    );
+
+    if (errors.length > 0) {
+      throw errors;
+    }
+
+    return this._convertToFixed(this.expressionCode.evaluate(historicalPrices));
+  }
+
+  getLastUpdateTime() {
+    const lastUpdateTimes = Object.values(this.priceFeedMap).map(pf => pf.getLastUpdateTime());
+
+    // If any constituents returned an invalid value, bubble it up.
+    if (lastUpdateTimes.some(time => time === null || time === undefined)) {
+      return null;
+    }
+
+    // Take the max.
+    return Math.max(...lastUpdateTimes);
+  }
+
+  getLookback() {
+    const lookbacks = Object.values(this.priceFeedMap).map(priceFeed => priceFeed.getLookback());
+
+    // If any constituents returned an invalid value, bubble it up.
+    if (lookbacks.some(lookback => lookback === undefined || lookback === null)) {
+      return null;
+    }
+
+    // Take the min since the overall lookback will be the min.
+    return Math.min(...lookbacks);
+  }
+
+  getCurrentPrice() {
+    const prices = {};
+    const errors = [];
+    Object.entries(this.priceFeedMap).map(([name, pf]) => {
+      try {
+        const price = pf.getCurrentPrice();
+        assert(price !== undefined && price !== null, "Valid price must be returned");
+        prices[name] = this._convertToDecimal(price, pf.getPriceFeedDecimals());
+      } catch (err) {
+        errors.push(err);
+      }
+    });
+
+    if (errors.length > 0) return null;
+
+    return this._convertToFixed(this.expressionCode.evaluate(prices));
```

Comment:
wonder if we can make this DRYer since the logic is really similar to getting historical price. its almost like you could create a higher order function which takes in a `getPrice` call which can represent getCurrentPrice or getHistorical price. wdyt?

Reply:
Two issues I see with that approach (the first likely making this a nonstarter):
- getHistoricalPrice is async and this higher level function would likely need to be async as a result of calling getHistoricalPrice on other feeds. If this function is async, getCurrentPrice can't call it. _Technically_ there may be ways around this by trying to do the async stuff outside of the function, but I think that might be so complex as to defeat the purpose of this refactor.
- I suspect this may hurt readability. I'm not sure, though (I'd have to try to implement it to really tell).

Thoughts?

Reply:
yes the sync/async issue does break that idea! thanks for thinking about this